shell/platform/darwin/ios/framework/Source/FlutterMetalLayer.mm

```diff
@@ -0,0 +1,271 @@
+    (id)kIOSurfaceBytesPerRow : @(bytesPerRow),
+    (id)kIOSurfaceAllocSize : @(totalBytes),
+  };
+
+  IOSurfaceRef res = IOSurfaceCreate((CFDictionaryRef)options);
+  if (self.colorspace != nil) {
+    CFStringRef name = CGColorSpaceGetName(self.colorspace);
+    IOSurfaceSetValue(res, CFSTR("IOSurfaceColorSpace"), name);
+  } else {
+    IOSurfaceSetValue(res, CFSTR("IOSurfaceColorSpace"), kCGColorSpaceSRGB);
+  }
+  return (__bridge_transfer IOSurface*)res;
+}
+
+- (id<CAMetalDrawable>)nextDrawable {
+  @synchronized(self) {
+    // IOSurface.isInUse to determine when compositor is done with the surface.
+    // That a rather blunt instrument and we are probably waiting longer than
+    // we really need to. With triple buffering at 120Hz that results in about
+    // 2-3 milliseconds wait time at beginning of display link callback.
+    // With four buffers this number gets close to zero.
+    if (_totalDrawables < 4) {
+      ++_totalDrawables;
+      IOSurface* surface = [self createIOSurface];
+      MTLTextureDescriptor* textureDescriptor =
+          [MTLTextureDescriptor texture2DDescriptorWithPixelFormat:_pixelFormat
+                                                             width:_drawableSize.width
+                                                            height:_drawableSize.height
+                                                         mipmapped:NO];
+      textureDescriptor.usage =
+          MTLTextureUsageShaderRead | MTLTextureUsageRenderTarget | MTLTextureUsageShaderWrite;
+      id<MTLTexture> texture = [self.device newTextureWithDescriptor:textureDescriptor
+                                                           iosurface:(__bridge IOSurfaceRef)surface
+                                                               plane:0];
+      FlutterDrawable* drawable = [[FlutterDrawable alloc] initWithTexture:texture
+                                                                     layer:self
+                                                                drawableId:_nextDrawableId++
+                                                                   surface:surface];
+      return drawable;
+    } else {
+      FlutterDrawable* res;
+      CFTimeInterval start = CACurrentMediaTime();
+
+      while (true) {
+        for (FlutterDrawable* drawable in _availableDrawables) {
+          if (!drawable.surface.inUse) {
+            res = drawable;
+            [_availableDrawables removeObject:drawable];
+            goto done;
+          }
+        }
+        usleep(10);
+      }
+    done:
+      CFTimeInterval duration = CACurrentMediaTime() - start;
+      if (duration > 0.003) {
+        NSLog(@"Getting drawable took %f", duration);
+      }
+      return res;
+    }
+  }
+  return nil;
+}
+
+- (void)presentOnMainThread:(FlutterDrawable*)drawable {
+  // This is needed otherwise frame gets skipped on touch begin / end. Go figure.
+  // Might also be placebo
+  [self setNeedsDisplay];
```

Comment:
Does this need to be fixed in our existing CAtransaction code?

Reply:
err, DisplayLink

Reply:
If you refering to `[self setNeedsDisplay] probably makes no difference for metal layer. I'm not even sure it makes difference here. It seems to have helped with frameskip on touch with 120hz display locked to 60hz, but that was several prototypes ago, on iOS 16 and without constant CADisplayLink on main thread. Thinking of it, the behavior might have been affected by touch rate correction displaylink (which is on main thread and enabled only during touch). I left it there just in case, but might be just placebo and needs to be retested.

Reply:
This will cause `FlutterView` to take a screenshot...

Reply:
It was removed from the final PR, but since the layer is backed by IOSurface I don't think it has associated `CGContextRef`, so I'd assume `[FlutterView drawLayer:inContext:]` should receive a nil context? Maybe I'm wrong about this.